### ISSUE TITLE
`needless_collect` turns `collect::<Vec<_>>()[..]` into `nth(..).unwrap()` turning our slice-cast into an failed indexing attempt

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#16270]
Using the following flags

`--force-warn clippy::needless-collect`

this code:

```rust
use std::{cmp, mem};

/// Finds the [edit distance] between two strings.
///
/// Returns `None` if the distance exceeds the limit.
///
/// [edit distance]: https://en.wikipedia.org/wiki/Edit_distance
pub fn edit_distance(a: &str, b: &str, limit: usize) -> Option<usize> {
    // Comparing the strings lowercased will result in a difference in capitalization being less distance away
    // than being a completely different letter. Otherwise `CHECK` is as far away from `check` as it
    // is from `build` (both with a distance of 5). For a single letter shortcut (e.g. `b` or `c`), they will
    // all be as far away from any capital single letter entry (all with a distance of 1).
    // By first lowercasing the strings, `C` and `c` are closer than `C` and `b`, for example.
    let a = a.to_lowercase();
    let b = b.to_lowercase();

    let mut a = &a.chars().collect::<Vec<_>>()[..];
    let mut b = &b.chars().collect::<Vec<_>>()[..];

    // Ensure that `b` is the shorter string, minimizing memory use.
    if a.len() < b.len() {
        mem::swap(&mut a, &mut b);
    }

    todo!();
}

```

caused the following diagnostics:

```
    Checking _a v0.1.0 (/tmp/icemaker_global_tempdir.BbwStV4YQFtQ/icemaker_clippyfix_tempdir.IBgjPBQjorLq/_a)
warning: avoid using `collect()` when not needed
  --> src/lib.rs:17:28
   |
17 |     let mut a = &a.chars().collect::<Vec<_>>()[..];
   |                            ^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `nth(..).unwrap()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_collect
   = note: requested on the command line with `--force-warn clippy::needless-collect`

warning: avoid using `collect()` when not needed
  --> src/lib.rs:18:28
   |
18 |     let mut b = &b.chars().collect::<Vec<_>>()[..];
   |                            ^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `nth(..).unwrap()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_collect

warning: `_a` (lib) generated 2 warnings
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.22s

```

However after applying these diagnostics, the resulting code:

```rust
use std::{cmp, mem};

/// Finds the [edit distance] between two strings.
///
/// Returns `None` if the distance exceeds the limit.
///
/// [edit distance]: https://en.wikipedia.org/wiki/Edit_distance
pub fn edit_distance(a: &str, b: &str, limit: usize) -> Option<usize> {
    // Comparing the strings lowercased will result in a difference in capitalization being less distance away
    // than being a completely different letter. Otherwise `CHECK` is as far away from `check` as it
    // is from `build` (both with a distance of 5). For a single letter shortcut (e.g. `b` or `c`), they will
    // all be as far away from any capital single letter entry (all with a distance of 1).
    // By first lowercasing the strings, `C` and `c` are closer than `C` and `b`, for example.
    let a = a.to_lowercase();
    let b = b.to_lowercase();

    let mut a = &a.chars().nth(..).unwrap();
    let mut b = &b.chars().nth(..).unwrap();

    // Ensure that `b` is the shorter string, minimizing memory use.
    if a.len() < b.len() {
        mem::swap(&mut a, &mut b);
    }

    todo!();
}

```
no longer compiled:

```
    Checking _a v0.1.0 (/tmp/icemaker_global_tempdir.BbwStV4YQFtQ/icemaker_clippyfix_tempdir.IBgjPBQjorLq/_a)
error[E0308]: mismatched types
   --> src/lib.rs:17:32
    |
 17 |     let mut a = &a.chars().nth(..).unwrap();
    |                            --- ^^ expected `usize`, found `RangeFull`
    |                            |
    |                            arguments to this method are incorrect
    |
note: method defined here
   --> /home/matthias/.rustup/toolchains/master/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:374:8
    |
374 |     fn nth(&mut self, n: usize) -> Option<Self::Item> {
    |        ^^^
help: you might have meant to use `..` to skip providing a value for expected fields, but this is only supported on non-tuple struct literals when `#![feature(default_field_values)]` is enabled; it is instead interpreted as a `std::ops::RangeFull` literal
   --> src/lib.rs:17:32
    |
 17 |     let mut a = &a.chars().nth(..).unwrap();
    |                                ^^

error[E0308]: mismatched types
   --> src/lib.rs:18:32
    |
 18 |     let mut b = &b.chars().nth(..).unwrap();
    |                            --- ^^ expected `usize`, found `RangeFull`
    |                            |
    |                            arguments to this method are incorrect
    |
note: method defined here
   --> /home/matthias/.rustup/toolchains/master/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:374:8
    |
374 |     fn nth(&mut self, n: usize) -> Option<Self::Item> {
    |        ^^^
help: you might have meant to use `..` to skip providing a value for expected fields, but this is only supported on non-tuple struct literals when `#![feature(default_field_values)]` is enabled; it is instead interpreted as a `std::ops::RangeFull` literal
   --> src/lib.rs:18:32
    |
 18 |     let mut b = &b.chars().nth(..).unwrap();
    |                                ^^

error[E0599]: no method named `len` found for reference `&char` in the current scope
    --> src/lib.rs:21:10
     |
  21 |     if a.len() < b.len() {
     |          ^^^
     |
note: there's an earlier shadowed binding `a` of type `std::string::String` that has method `len` available
    --> src/lib.rs:14:9
     |
  14 |     let a = a.to_lowercase();
     |         ^ `a` of type `std::string::String` that has method `len` defined earlier here
...
  17 |     let mut a = &a.chars().nth(..).unwrap();
     |         ----- earlier `a` shadowed here with type `&char`
help: there is a method `le` with a similar name, but with different arguments
    --> /home/matthias/.rustup/toolchains/master/lib/rustlib/src/rust/library/core/src/cmp.rs:1416:5
     |
1416 |     fn le(&self, other: &Rhs) -> bool {
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0599]: no method named `len` found for reference `&char` in the current scope
    --> src/lib.rs:21:20
     |
  21 |     if a.len() < b.len() {
     |                    ^^^
     |
note: there's an earlier shadowed binding `b` of type `std::string::String` that has method `len` available
    --> src/lib.rs:15:9
     |
  15 |     let b = b.to_lowercase();
     |         ^ `b` of type `std::string::String` that has method `len` defined earlier here
...
  18 |     let mut b = &b.chars().nth(..).unwrap();
     |         ----- earlier `b` shadowed here with type `&char`
help: there is a method `le` with a similar name, but with different arguments
    --> /home/matthias/.rustup/toolchains/master/lib/rustlib/src/rust/library/core/src/cmp.rs:1416:5
     |
1416 |     fn le(&self, other: &Rhs) -> bool {
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Some errors have detailed explanations: E0308, E0599.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `_a` (lib test) due to 4 previous errors
warning: build failed, waiting for other jobs to finish...
error: could not compile `_a` (lib) due to 4 previous errors

```

Version:
```
rustc 1.94.0-nightly (f51d1bcdc 2025-12-20)
binary: rustc
commit-hash: f51d1bcdc69f9bba47cc1656260ff76e6ff5fba3
commit-date: 2025-12-20
host: x86_64-unknown-linux-gnu
release: 1.94.0-nightly
LLVM version: 21.1.8

```
